homeassistant/components/mqtt/valve.py

```diff
@@ -0,0 +1,386 @@
+        vol.Optional(CONF_STATE_CLOSING, default=STATE_CLOSING): cv.string,
+        vol.Optional(CONF_STATE_OPEN): cv.string,
+        vol.Optional(CONF_STATE_OPENING, default=STATE_OPENING): cv.string,
+        vol.Optional(CONF_STATE_TOPIC): valid_subscribe_topic,
+        vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
+    }
+).extend(MQTT_ENTITY_COMMON_SCHEMA.schema)
+
+PLATFORM_SCHEMA_MODERN = vol.All(_PLATFORM_SCHEMA_BASE, _validate_valve_payload_options)
+
+DISCOVERY_SCHEMA = vol.All(
+    _PLATFORM_SCHEMA_BASE.extend({}, extra=vol.REMOVE_EXTRA),
+    _validate_valve_payload_options,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up MQTT valve through YAML and through MQTT discovery."""
+    await async_setup_entity_entry_helper(
+        hass,
+        config_entry,
+        MqttValve,
+        valve.DOMAIN,
+        async_add_entities,
+        DISCOVERY_SCHEMA,
+        PLATFORM_SCHEMA_MODERN,
+    )
+
+
+class MqttValve(MqttEntity, ValveEntity):
+    """Representation of a valve that can be controlled using MQTT."""
+
+    _attr_is_closed: bool | None = None
+    _attributes_extra_blocked: frozenset[str] = MQTT_VALVE_ATTRIBUTES_BLOCKED
+    _default_name = DEFAULT_NAME
+    _entity_id_format: str = valve.ENTITY_ID_FORMAT
+    _optimistic: bool
+    _range: tuple[int, int]
+    _tilt_optimistic: bool
+
+    @staticmethod
+    def config_schema() -> vol.Schema:
+        """Return the config schema."""
+        return DISCOVERY_SCHEMA
+
+    def _setup_from_config(self, config: ConfigType) -> None:
+        """Set up valve from config."""
+        self._attr_reports_position = config[CONF_REPORTS_POSITION]
+        self._range = (
+            self._config[CONF_POSITION_CLOSED] + 1,
+            self._config[CONF_POSITION_OPEN],
+        )
+        no_state_topic = config.get(CONF_STATE_TOPIC) is None
+        self._optimistic = config[CONF_OPTIMISTIC] or no_state_topic
+        self._attr_assumed_state = self._optimistic
+
+        template_config_attributes = {
+            "position_open": config[CONF_POSITION_OPEN],
+            "position_closed": config[CONF_POSITION_CLOSED],
+        }
+
+        self._value_template = MqttValueTemplate(
+            config.get(CONF_VALUE_TEMPLATE), entity=self
+        ).async_render_with_possible_json_value
+
+        self._command_template = MqttCommandTemplate(
+            config.get(CONF_COMMAND_TEMPLATE), entity=self
+        ).async_render
+
+        self._value_template = MqttValueTemplate(
+            config.get(CONF_VALUE_TEMPLATE),
+            entity=self,
+            config_attributes=template_config_attributes,
+        ).async_render_with_possible_json_value
+
+        self._attr_device_class = config.get(CONF_DEVICE_CLASS)
+
+        supported_features = ValveEntityFeature(0)
+        if has_command_topic := (config.get(CONF_COMMAND_TOPIC) is not None):
```

Comment:
This seems backwards, isn't this clearer:
```suggestion
        if has_command_topic := (CONF_COMMAND_TOPIC in config):
```

If yes, please make the same change also below